internal/grpcutil/batch.go

```diff
@@ -0,0 +1,85 @@
+package grpcutil
+
+import (
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+type batch struct {
+	no    int
+	start int
+	end   int
+}
+
+func worker(errs chan<- error, queue <-chan batch, each EachFunc) {
+	for span := range queue {
+		err := each(span.no, span.start, span.end)
+		errs <- errors.Wrapf(err, "batch %d failed", span.no)
+	}
+}
+
+// EachFunc is a callback function that is called for each batch. no is the
+// batch number, start is the starting index of this batch in the slice, and
+// end is the ending index of this batch in the slice.
+type EachFunc func(no int, start int, end int) error
+
+// Batch will calculate the minimum number of batches to required to batch n
+// items with batchSize batches. For each batch, it will execute the each
+// function. These functions will be processed in parallel using workers number
+// of goroutines.
+func Batch(n int, batchSize int, workers int, each EachFunc) error {
+	if n < 0 {
+		return errors.New("cannot batch items of length < 0")
+	}
+
+	if batchSize < 1 {
+		return errors.New("cannot batch items with batch size < 1")
+	}
+
+	if workers < 1 {
+		return errors.New("cannot batch items with workers < 1")
+	}
+
+	if n == 0 {
+		// Batching zero items is a noop.
+		return nil
+	}
+
+	totalBatches := (n + batchSize - 1) / batchSize
+
+	// Queue up all the batches.
+	batches := make(chan batch, totalBatches)
+	for i := 0; i < totalBatches; i++ {
+		start := i * batchSize
+		end := start + batchSize
+		if end > n {
+			end = n
+		}
+
+		batches <- batch{i + 1, start, end}
+	}
+	close(batches)
+
+	// Start the workers.
+	errCh := make(chan error)
+	for i := 0; i < workers; i++ {
+		go worker(errCh, batches, each)
+	}
+
+	// Wait until all the workers have finished processing the queue.
+	var err *multierror.Error
+	var finishedBatches int
+	for e := range errCh {
+		finishedBatches++
+
+		if e != nil {
+			err = multierror.Append(err, e)
+		}
+
+		if finishedBatches == totalBatches {
+			close(errCh)
+		}
+	}
+
+	return err.ErrorOrNil()
+}
```

Comment:
It's a bit of personal preference, but I like to follow the lead for the x/sync packages when it comes to implementing concurrency patterns.

You can find a WorkerPool example here: https://pkg.go.dev/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/semaphore

I haven't tested this, but something along the lines of this would be preferable and wouldn't add any new dependencies:

```go
func min(x, y int) int {
	if x <= y {
		return x
	}
	return y
}

func ConcurrentBatch(ctx context.Context, n, batchSize, maxWorkers int, each EachFunc) error {
	if n < 0 {
		return errors.New("cannot batch items of length < 0")
	} else if n == 0 {
		return nil // Batching zero items is a noop.
	}

	if batchSize < 1 {
		return errors.New("cannot batch items with batch size < 1")
	}

	if maxWorkers == 0 {
		maxWorkers = runtime.GOMAXPROCS(0)
	}

	sem := semaphore.NewWeighted(int64(maxWorkers))
	g, ctx := errgroup.WithContext(ctx)
	numBatches := (n + batchSize - 1) / batchSize
	for i := 0; i < numBatches; i++ {
		if err := sem.Acquire(ctx, 1); err != nil {
			return fmt.Errorf("failed to acquire semaphore for batch number %d: %w", i, err)
		}
		g.Go(func() error {
			defer sem.Release(1)
			batchNum := i // Properly copy i
			start := i * batchSize
			end := min(start+batchSize, n)
			return each(batchNum, start, end)
		})
	}
	return g.Wait()
}
```

The reason I use these primitives instead is because channels actually have a ton of edge cases. You can see some in this paper that might surprise you, and I've found others in production systems: https://cseweb.ucsd.edu/~yiying/GoStudy-ASPLOS19.pdf

Reply:
That paper is an extremely interesting read with a few surprising "gotchas" when it comes to channels. I'll adopt the `WorkerPool` approach. I also like the name being changed to `ConcurrentBatch`, indicating that things happen concurrently under the hood.